### ISSUE TITLE
Find a way to distinguish regular users from bots 

### ISSUE THREAD

**Issue by zurk** [src-d/identity-matching#9]
We can take some rule-based approach as a benchmark: email contains `bot` word or `no-reply`. However, there are emails like `tensorflow-gardener@tensorflow.org` that is hard to find. So some ML should be applied to find them. Commit-time-series features can be used.

**Comment by EgorBu** [comment]
### There are at least several problems that may affect the quality:
1. Noisy labels - 
    * false positives from regexp - like: `abbot`, `julia jenkins` and so on
    * false negatives - not detected bots (`gardener@tensorflow` for example)
2. Model input doesn't contain required info to make a correct prediction
    * false negatives - email doesn't contain bot related info, and the name contains. Ex: `egors-bot-deploy@bla.ru / Egorka`
3. The name doesn't contain the required information to label it as bot
    * false positives - email contains bot related info, and the name doesn't contain. Ex: `egors-bot-deploy@bla.ru / Egorka` -> so it will be labeled as not bot and email tells that it's a bot
4. Metrics. Deduplication:
    * deduplication is done by several fields - and if `repository` name is included - the quality could be found here - https://gist.github.com/EgorBu/a333409dfc12f89ac5fa1dc71461a3c0 
    * it's higher than current - probably it could mean that standard names for bots are much more frequent - and in most of the cases standard names will be detected with high quality
5. Metrics. Usage
    * we still don't have a clear understanding of how it should be applied (for each commit, for each identity, etc) - metrics should be selected on usage
6. Dataset
    * Another possible reason that quality was higher [here](https://gist.github.com/EgorBu/a333409dfc12f89ac5fa1dc71461a3c0 ) is some dataset issues

### Hypothesis to check
* Dataset
    * select a row in dataset
    * split dataset into 2 parts before some row and after
    * assign labels (0 - before, 1 - after)
    * train classifier - if the quality is better than random - something is fishy with dataset
*  `false positives - email contains bot related info, and the name doesn't contain` and `false negatives - email doesn't contain bot related info, and the name contains`
    * labels & predictions should be computed
    * extract features separately from names & emails
    * find nearest neighbors by name
    * find nearest neighbors by email
    * several situations are possible:
        * labels & predictions are the same among nearest neighbors for name & emails - perfect
        * labels among nearest neighbors for name are not the same - possible regexp mistakes?
        * predictions are not the same among nearest neighbors for emails - check it
        * labels & predictions are not the same among nearest neighbors for name & emails - possible regexp mistake?
* model overfits to mistakes on regexp
    * hypothesis - number of mistakes is not so big
    * train several models on different chunks of data - it will reduce number of mistakes in each chunk
    * make voting among models when making prediction
    * focus on samples with different predictions and labels
    * focus on samples with different predictions
* features are not good enough
    * BPE could extract features from `abot` as `[a, bot]` - and it will make almost impossible for model to differentiate one class from another
        * use `token splitter` to split `victor@abot.fr` into `[victor, abot, fr]`
        * add a feature that will highlight if something is in the exception list
        * don't extract BPE features from exceptions
